src/global/pidfile.cc

```diff
@@ -31,68 +31,131 @@
-    return pidfile_remove();
-  }
-  snprintf(pid_file, PATH_MAX, "%s", conf->pid_file.c_str());
+     if (pfh.pf_fd == -1)
+            return -EINVAL;
+     /*
+      * Check remembered descriptor
+      */ 
+    if (fstat(pfh.pf_fd, &sb) == -1)
+            return -errno;
+    if (sb.st_dev != pfh.pf_dev || sb.st_ino != pfh.pf_ino)
+           return -ESTALE;
+    return 0;
+}
 
-  fd = TEMP_FAILURE_RETRY(::open(pid_file,
-				 O_CREAT|O_TRUNC|O_WRONLY, 0644));
-  if (fd < 0) {
-    int err = errno;
-    derr << "write_pid_file: failed to open pid file '"
-	 << pid_file << "': " << cpp_strerror(err) << dendl;
-    return err;
-  }
+int pidfile_write()
+{
+  int ret;
+  if (!pfh.pf_path[0])
+      return 0;
 
-  char buf[20];
+  if ((ret=pidfile_verify()) < 0) {
+    return ret;
+  }
+  char buf[32];
   int len = snprintf(buf, sizeof(buf), "%d\n", getpid());
-  ret = safe_write(fd, buf, len);
+  ret = safe_write(pfh.pf_fd, buf, len);
   if (ret < 0) {
     derr << "write_pid_file: failed to write to pid file '"
-	 << pid_file << "': " << cpp_strerror(ret) << dendl;
-    VOID_TEMP_FAILURE_RETRY(::close(fd));
+	 << pfh.pf_path << "': " << cpp_strerror(ret) << dendl;
+    pfh.close(); 
     return ret;
   }
-  if (TEMP_FAILURE_RETRY(::close(fd))) {
-    ret = errno;
-    derr << "SimpleMessenger::write_pid_file: failed to close to pid file '"
-	 << pid_file << "': " << cpp_strerror(ret) << dendl;
-    return -ret;
-  }
 
   return 0;
 }
 
 int pidfile_remove(void)
 {
-  if (!pid_file[0])
+  int ret;
+ 
+  if (!pfh.pf_path[0])
     return 0;
+  if ( (ret=pidfile_verify()) < 0) {
+      VOID_TEMP_FAILURE_RETRY(::close(pfh.pf_fd));
+      return ret;
+  }
 
   // only remove it if it has OUR pid in it!
-  int fd = TEMP_FAILURE_RETRY(::open(pid_file, O_RDONLY));
-  if (fd < 0)
-    return -errno;
   char buf[32];
   memset(buf, 0, sizeof(buf));
-  ssize_t res = safe_read(fd, buf, sizeof(buf));
-  VOID_TEMP_FAILURE_RETRY(::close(fd));
+  ssize_t res = safe_read(pfh.pf_fd, buf, sizeof(buf));
+  VOID_TEMP_FAILURE_RETRY(::close(pfh.pf_fd));
   if (res < 0)
-    return res;
+     return res;
   int a = atoi(buf);
   if (a != getpid())
     return -EDOM;
-
-  res = ::unlink(pid_file);
+  
+  res = ::unlink(pfh.pf_path);
   if (res)
-    return res;
-
-  pid_file[0] = '\0';
+     return res;
+  pfh.close();
   return 0;
 }
+
+int pidfile_open(const md_config_t *conf) 
+{
+  int fd;
+  struct stat sb;
+  if (conf->pid_file.empty()) {
+    return 0;
+  }
+  int len = snprintf(pfh.pf_path, sizeof(pfh.pf_path),
+                    "%s", conf->pid_file.c_str());
+
+  if (len >= (int)sizeof(pfh.pf_path)) { 
+      return -ENAMETOOLONG;
+  }
+ 
+  fd = ::open(pfh.pf_path, O_CREAT|O_WRONLY, 0644);
+  if (fd < 0) {
+     int err = errno;
+     derr << "write_pid_file: failed to open pid file '"
+          << pfh.pf_path << "': " << cpp_strerror(err) << dendl;
+     pfh.close();
+     return -errno;
+   }
+    
+   if (fstat(fd, &sb) == -1) {
```

Comment:
What do you mean? Returning -1 and setting errno is POSIX standard, do you know any system that behaves differently? I like checking for -1 more because, when I see '< 0' I expect the function returns -error, which is not this case.